pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -418,6 +446,42 @@ func (b *BinaryOperation) mergeOneSide(data []types.InstantVectorSeriesData, sou
 		return types.InstantVectorSeriesData{}, nil
 	}
 
+	// Merge floats and histograms individually.
+	// After which we check if there are any duplicate points in either the floats or histograms.
+	filteredData := filteredData{}
+	filteredData.filter(data)
+
+	floats, err := b.mergeOneSideFloats(filteredData, sourceSeriesIndices, sourceSeriesMetadata, side)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+	histograms, err := b.mergeOneSideHistograms(filteredData, sourceSeriesIndices, sourceSeriesMetadata, side)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	// Check for any conflicts between floats and histograms
+	i, j := 0, 0
+	for i < len(floats) && j < len(histograms) {
+		if floats[i].T == histograms[j].T {
+			// Conflict found
+			return types.InstantVectorSeriesData{}, fmt.Errorf("found both float and histogram samples for the match group FIXME on the %s side of the operation at timestamp %s", side, timestamp.Time(floats[i].T))
+		}
+		if floats[i].T < histograms[j].T {
+			i++
+		} else {
+			j++
+		}
+	}
+
+	return types.InstantVectorSeriesData{Floats: floats, Histograms: histograms}, nil
+}
+
+func (b *BinaryOperation) mergeOneSideFloats(filteredData filteredData, sourceSeriesIndices []int, sourceSeriesMetadata []types.SeriesMetadata, side string) ([]promql.FPoint, error) {
+	data := filteredData.dataWithFloats
+	if len(data) == 0 {
+		return nil, nil
+	}
 	slices.SortFunc(data, func(a, b types.InstantVectorSeriesData) int {
 		return int(a.Floats[0].T - b.Floats[0].T)
 	})
```

Comment:
Just realised this was not correct previously, and the problem still exists after your changes: if we reorder series here, we don't reorder the corresponding indices in `sourceSeriesIndices` at the same time, so the conflict error message might refer to incorrect series names.

For example, imagine we have three input series, in this order:

S1: samples from T=5 to T=10
S2: samples from T=0 to T=4
S3: samples from T=6 to T=10, with a conflict with S1

We'll sort these to [S2, S1, S3], but when we produce the conflict error message below, we'll use the labels of S2 and S3.

We might need to use something similar to `groupSorter` from the aggregation operator so we sort and swap elements in both slices at the same time.